lib/ts/recipe/webauthn/types.ts

```diff
@@ -0,0 +1,747 @@
+            timeout: number | undefined;
+            tenantId: string;
+            userContext: UserContext;
+        } & (
+            | {
+                  recoverAccountToken: string;
+              }
+            | {
+                  email: string;
+              }
+        )
+    ): Promise<
+        | {
+              status: "OK";
+              webauthnGeneratedOptionsId: string;
+              rp: {
+                  id: string;
+                  name: string;
+              };
+              user: {
+                  id: string;
+                  name: string; // user email
+                  displayName: string; //user email
+              };
+              challenge: string;
+              timeout: number;
+              excludeCredentials: {
+                  id: string;
+                  type: "public-key";
+                  transports: ("ble" | "hybrid" | "internal" | "nfc" | "usb")[];
+              }[];
+              attestation: "none" | "indirect" | "direct" | "enterprise";
+              pubKeyCredParams: {
+                  // we will default to [-8, -7, -257] as supported algorithms. See https://www.iana.org/assignments/cose/cose.xhtml#algorithms
+                  alg: number;
+                  type: "public-key";
+              }[];
+              authenticatorSelection: {
+                  requireResidentKey: boolean;
+                  residentKey: "required" | "preferred" | "discouraged";
+                  userVerification: "required" | "preferred" | "discouraged";
+              };
+          }
+        // | RegisterOptionsErrorResponse
+        | { status: "RECOVER_ACCOUNT_TOKEN_INVALID_ERROR" }
+        | { status: "INVALID_EMAIL_ERROR"; err: string }
+    >;
+
+    signInOptions(input: {
+        email?: string;
+        relyingPartyId: string;
+        origin: string;
+        userVerification: "required" | "preferred" | "discouraged" | undefined; // see register options
+        timeout: number | undefined;
+        tenantId: string;
+        userContext: UserContext;
+    }): Promise<
+        | {
+              status: "OK";
+              webauthnGeneratedOptionsId: string;
+              challenge: string;
+              timeout: number;
+              userVerification: "required" | "preferred" | "discouraged";
+          }
+        // | SignInOptionsErrorResponse
+        | { status: "WRONG_CREDENTIALS_ERROR" }
+    >;
+
+    signUp(input: {
+        webauthnGeneratedOptionsId: string;
+        credential: CredentialPayload;
+        session: SessionContainerInterface | undefined;
+        shouldTryLinkingWithSessionUser: boolean | undefined;
+        tenantId: string;
+        userContext: UserContext;
+    }): Promise<
+        | {
+              status: "OK";
+              user: User;
+              recipeUserId: RecipeUserId;
+          }
+        // | SignUpErrorResponse
+        | { status: "EMAIL_ALREADY_EXISTS_ERROR" }
+        | { status: "WRONG_CREDENTIALS_ERROR" }
```

Comment:
This feels weird on a `signUp` endpoint, maybe specify which param is wrong/how?

Reply:
The status is a generic one, for whatever error happens because of the Webauthn credentials flow. If we'd need to be more specific here (not in the API implementation - we should keep it generic there), then we'd need to do the actual core implementation, as there might lots of stuff that we can only find out when we start implementing.

There is another type of status "INVALID_AUTHENTICATOR_ERROR" (along with the reason) which means that the credential has been generated by an "unsupported" authenticator (this could happen by overriding the registerOptions method and overriding the default values for `attestation`, `requireResidentKey`, `residentKey`, `userVerification`, `supportedAlgorithmIds`, etc.).

What do you think ?

Reply:
The `INVALID_AUTHENTICATOR_ERROR` sounds good. My main issue with `WRONG_CREDENTIALS_ERROR` is that it overlaps with the error status we have for people submitting wrong passwords I guess.

Reply:
same as previous error discussion

Reply:
done